mailz.1
.Dd $Mdocdate: Mar 25 2025$
.Dt MAILZ 1
.Os
.Sh NAME
.Nm mailz
.Nd view and reply to email
.Sh SYNOPSIS
.Nm mailz
.Op Fl a
.Op Fl f Ar file
.Op Fl n
.Ar mailbox
.Sh DESCRIPTION
The
.Nm
utility provides a command line interface to interact with mail located
in a directory.
The
.Ar mailbox
operand names a directory storing mail in the maildir format, or a
mailbox as specified in
.Xr mailz.conf 5 .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
Display mail that has already been read.
.It Fl f Ar file
Specifies the configuration file.
The default is
.Pa $HOME/.mailz.conf .
.It Fl n
Check that the configuration file is valid and exit.
.El
.Pp
Upon startup,
.Nm
will display a listing of all mail received.
Commands can then be entered in a
.Xr sh 1
like interface.
Command arguments are separated by space or tab characters,
and each command is terminated by a newline.
Each command accepts zero or more message numbers to operate on.
If a message number is not provided, the last message interacted with
will be used (if any).
If a message number is prefixed by 't' followed by a space
each message in the same thread as that message will be operated
on.
.Pp
The available commands are as follows:
.Bl -tag -width Ds
.It Ic delete
Mark each message as deleted.
Does not delete the on-disk file.
.It Ic more
Open each message in the
.Xr less 1
pager, and mark them as having been read.
.It Ic read (r)
Mark each message as having been read.
.It Ic reply
Reply to each message, addressing the sender and recipients of the
message.
A template reply is presented for editing before sending the reply.
.It Ic respond
Reply to each message, addressing the sender of the message.
A template reply is presented for editing before sending the reply.
.It Ic save (s)
Save each message to a temporary file and print its location.
.It Ic thread (t)
For each message, list all messages in the same thread.
.It Ic unread (x)
Mark each message as not having been read.
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa ~/.mailz.conf
Default configuration file.
.It Pa ~/.mailz/
Temporary directory used by
.Nm .
.It Pa ~/.mailz/reply.*
Messages to be sent by the
.Ic reply
command.
.It Pa ~/.mailz/save.*
Messages saved by the
.Ic save
command.
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, 1 on error, and 2 on invalid usage.
.Sh SEE ALSO
.Xr mailz.conf 5
.Sh AUTHORS
The
.Nm
utility was written by
.An Henry Ford Aq Mt fordhenry2299@gmail.com
.Sh BUGS
The
.Ic thread
command cannot cope with email threads where the subject is
changed, or with multiple email threads with the same subject.


mailz.c
/*
 * Copyright (c) 2024 Henry Ford <fordhenry2299@gmail.com>

 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.

 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/stat.h>
#include <sys/wait.h>

#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <locale.h>
#include <time.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "command.h"
#include "conf.h"
#include "content-proc.h"
#include "err-fork.h"
#include "mailbox.h"
#include "maildir.h"
#include "pathnames.h"

struct command_args {
	const char *addr;
	const char *maildir;
	const char *tmpdir;
	struct config_ignore *ignore;
	struct mailbox *mailbox;
	int cur;
};

#define nitems(a) (sizeof((a)) / sizeof(*(a)))

static void commands_run(struct command_args *);
static const struct command *commands_search(const char *);
static int command_delete(struct letter *, struct command_args *);
static int command_flag(struct letter *, struct command_args *, int,
			int);
static int command_more(struct letter *, struct command_args *);
static int command_read(struct letter *, struct command_args *);
static int command_reply1(struct letter *, struct command_args *, int);
static int command_reply(struct letter *, struct command_args *);
static int command_respond(struct letter *, struct command_args *);
static int command_save(struct letter *, struct command_args *);
static int command_thread(struct letter *, struct command_args *);
static int command_unread(struct letter *, struct command_args *);
static int content_proc_ex_ignore(struct content_proc *,
				  const struct config_ignore *);
static int letter_print(size_t, struct letter *);
static void read_letters(const char *, int, int, struct mailbox *);
static void setup_letters(const char *, int, int);
static void usage(void);

static const struct command {
	const char *ident;
	#define CMD_NOALIAS '\0'
	int alias;
	int (*fn) (struct letter *, struct command_args *);
} commands[] = {
	{ "delete",	CMD_NOALIAS,	command_delete },
	{ "more",	CMD_NOALIAS,	command_more },
	{ "read",	'r',		command_read },
	{ "reply",	CMD_NOALIAS,	command_reply },
	{ "respond",	CMD_NOALIAS,	command_respond },
	{ "save",	's',		command_save },
	{ "thread",	't',		command_thread },
	{ "unread",	'x',		command_unread },
};

static void
commands_run(struct command_args *args)
{
	struct command_lexer lex;
	struct letter *letter;

	command_init(&lex, stdin);
	letter = NULL;
	for (;;) {
		const struct command *cmd;
		char buf[8];
		int any, error;

		printf("> ");
		fflush(stdout);

		error = command_name(&lex, buf, sizeof(buf));
		if (error == COMMAND_EOF)
			break;
		if (error != COMMAND_OK) {
			switch (error) {
			case COMMAND_EMPTY:
				break;
			case COMMAND_LONG:
				warnx("command name too long");
				break;
			default:
				warnx("invalid command name");
				break;
			}

			continue;
		}

		if ((cmd = commands_search(buf)) == NULL) {
			warnx("unknown command");
			continue;
		}

		any = 0;
		for (;;) {
			struct command_letter cmd_letter;

			error = command_letter(&lex, &cmd_letter);
			if (error == COMMAND_EOF)
				break;
			any = 1;
			if (error != COMMAND_OK) {
				switch (error) {
				case COMMAND_INVALID:
					warnx("letter number invalid");
					break;
				case COMMAND_LONG:
					warnx("letter number too long");
					break;
				case COMMAND_THREAD_EOF:
					warnx("must provide a message number after 't'");
					break;
				default:
					break;
				}

				break;
			}

			/* These are numbered from 1, so no = */
			if (cmd_letter.num > args->mailbox->nletter) {
				warnx("letter number too large");
				break;
			}
			letter = &args->mailbox->letters[cmd_letter.num - 1];

			if (cmd_letter.thread) {
				struct letter *lp;
				struct mailbox_thread thread;

				mailbox_thread_init(args->mailbox, &thread, letter);
				while ((lp = mailbox_thread_next(args->mailbox, &thread)) != NULL) {
					if (cmd->fn(lp, args) == -1) {
						warnx("command '%s' failed", cmd->ident);
						goto next;
					}
				}
			}
			else {
				if (cmd->fn(letter, args) == -1) {
					warnx("command '%s' failed", cmd->ident);
					break;
				}
			}
		}

		if (!any) {
			if (letter == NULL) {
				warnx("no current letter");
				continue;
			}
			if (cmd->fn(letter, args) == -1)
				warnx("command '%s' failed", cmd->ident);
		}

		next:
		continue;
	}

	printf("\n");
}

static const struct command *
commands_search(const char *s)
{
	size_t i, len;

	len = strlen(s);
	for (i = 0; i < nitems(commands); i++) {
		if ((len == 1 && commands[i].alias == *s)
				|| !strcmp(s, commands[i].ident))
			return &commands[i];
	}
	return NULL;
}

static int
command_delete(struct letter *letter, struct command_args *args)
{
	return command_flag(letter, args, 'T', 1);
}

static int
command_flag(struct letter *letter, struct command_args *args,
	     int flag, int set)
{
	char buf[NAME_MAX + 1], *new;
	int error;

	if (set) {
		error = maildir_set_flag(letter->path, flag, buf,
					sizeof(buf));
	}
	else {
		error = maildir_unset_flag(letter->path, flag, buf,
					sizeof(buf));
	}

	if (error == MAILDIR_UNCHANGED)
		return 0;
	if (error != MAILDIR_OK) {
		switch (error) {
		case MAILDIR_INVALID:
			warnx("%s/cur/%s: invalid maildir info",
			      args->maildir, letter->path);
			break;
		case MAILDIR_LONG:
			warnx("%s/cur/%s: filename too long to modify",
			      args->maildir, letter->path);
			break;
		default: /* cant happen */
			warnx("%s/cur/%s", args->maildir, letter->path);
			break;
		}
		return -1;
	}

	if ((new = strdup(buf)) == NULL) {
		warn(NULL);
		return -1;
	}

	if (renameat(args->cur, letter->path, args->cur, new) == -1) {
		warn("rename %s/cur/%s to %s/cur/%s", args->maildir, letter->path,
		     args->maildir, new);
		free(new);
		return -1;
	}

	free(letter->path);
	letter->path = new;

	return 0;
}


static int
command_more(struct letter *letter, struct command_args *args)
{
	struct content_proc pr;
	struct content_letter lr;
	FILE *fp;
	int fd, p[2], rv;
	pid_t pid;

	rv = -1;

	if (content_proc_init(&pr, PATH_MAILZ_CONTENT) == -1)
		return -1;

	if (content_proc_ex_ignore(&pr, args->ignore) == -1)
		goto pr;

	if ((fd = openat(args->cur, letter->path,
			 O_RDONLY | O_CLOEXEC)) == -1)
		goto pr;
	if (content_letter_init(&pr, &lr, fd) == -1)
		goto pr;

	if (pipe2(p, O_CLOEXEC) == -1)
		goto lr;
	if ((fp = fdopen(p[1], "w")) == NULL) {
		close(p[1]);
		close(p[0]);
		goto lr;
	}

	switch (pid = fork()) {
	case -1:
		fclose(fp);
		close(p[0]);
		goto lr;
	case 0:
		if (dup2(p[0], STDIN_FILENO) == -1)
			err_fork(1, "dup2");
		execl(PATH_LESS, "less", NULL);
		err_fork(1, "%s", PATH_LESS);
	default:
		break;
	}
	close(p[0]);

	for (;;) {
		char buf[4];
		int n;

		if ((n = content_letter_getc(&lr, buf)) == -1)
			goto pid;
		if (n == 0)
			break;

		if (fwrite(buf, n, 1, fp) != 1) {
			if (ferror(fp) && errno == EPIPE)
				break;
			goto pid;
		}
	}

	if (command_read(letter, args) == -1)
		goto pid;

	rv = 0;
	pid:
	fclose(fp);
	waitpid(pid, NULL, 0);
	lr:
	content_letter_close(&lr);
	pr:
	content_proc_kill(&pr);
	return rv;
}

static int
command_read(struct letter *letter, struct command_args *args)
{
	return command_flag(letter, args, 'S', 1);
}

static int
command_reply(struct letter *letter, struct command_args *args)
{
	return command_reply1(letter, args, 1);
}

static int
command_reply1(struct letter *letter, struct command_args *args, int group)
{
	struct content_proc pr;
	char path[PATH_MAX];
	FILE *fp;
	pid_t pid;
	int ch, fd, lfd, n, rv, status;

	rv = -1;

	if (content_proc_init(&pr, PATH_MAILZ_CONTENT) == -1)
		return -1;

	n = snprintf(path, sizeof(path), "%s/reply.XXXXXX", args->tmpdir);
	if (n < 0 || (size_t)n >= sizeof(path))
		goto pr;

	if ((fd = mkostemp(path, O_CLOEXEC)) == -1)
		goto pr;
	if ((fp = fdopen(fd, "w")) == NULL) {
		unlink(path);
		close(fd);
		goto pr;
	}

	if ((lfd = openat(args->cur, letter->path, O_RDONLY | O_CLOEXEC)) == -1)
		goto fp;
	if (content_proc_reply(&pr, fp, args->addr, group, lfd) == -1)
		goto fp;

	if (fflush(fp) == EOF)
		goto fp;

	if (printf("message located at %s\n"
		   "press enter to send or q to cancel: ", path) < 0)
		goto fp;
	if (fflush(stdout) == EOF)
		goto fp;

	if ((ch = fgetc(stdin)) == EOF)
		goto fp;

	if (ch != '\n') {
		int any, c;

		any = 0;
		while ((c = fgetc(stdin)) == EOF && c != '\n')
			any = 1;
		if (ch == 'q' && !any)
			rv = 0;
		goto fp;
	}

	lfd = fileno(fp);
	if (lseek(lfd, 0, SEEK_SET) == -1)
		goto fp;
	switch (pid = fork()) {
	case -1:
		goto fp;
	case 0:
		if (dup2(lfd, STDIN_FILENO) == -1)
			err_fork(1, "dup2");
		execl(PATH_SENDMAIL, "sendmail", "-t", NULL);
		err_fork(1, "%s", PATH_SENDMAIL);
	default:
		break;
	}

	if (waitpid(pid, &status, 0) == -1)
		goto fp;
	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0)
		goto fp;

	rv = 0;
	fp:
	fclose(fp);
	unlink(path);
	pr:
	content_proc_kill(&pr);
	return rv;
}

static int
command_respond(struct letter *letter, struct command_args *args)
{
	return command_reply1(letter, args, 0);
}

static int
command_save(struct letter *letter, struct command_args *args)
{
	struct content_proc pr;
	struct content_letter lr;
	char path[PATH_MAX];
	FILE *fp;
	int fd, lfd, n, rv;

	rv = -1;

	if (content_proc_init(&pr, PATH_MAILZ_CONTENT) == -1)
		return -1;

	if (content_proc_ex_ignore(&pr, args->ignore) == -1)
		goto pr;

	if ((lfd = openat(args->cur, letter->path,
			  O_RDONLY | O_CLOEXEC)) == -1)
		goto pr;
	if (content_letter_init(&pr, &lr, lfd) == -1)
		goto pr;

	n = snprintf(path, sizeof(path), "%s/save.XXXXXX", args->tmpdir);
	if (n < 0 || (size_t)n >= sizeof(path))
		goto lr;

	if ((fd = mkostemp(path, O_CLOEXEC)) == -1)
		goto lr;
	if ((fp = fdopen(fd, "w")) == NULL) {
		unlink(path);
		close(fd);
		goto lr;
	}

	for (;;) {
		char buf[4];
		int n;

		if ((n = content_letter_getc(&lr, buf)) == -1)
			goto fp;
		if (n == 0)
			break;

		if (fwrite(buf, n, 1, fp) != 1)
			goto fp;
	}

	if (fflush(fp) == EOF)
		goto fp;

	if (printf("message saved to %s\n", path) < 0)
		goto fp;

	rv = 0;
	fp:
	fclose(fp);
	if (rv == -1)
		unlink(path);
	lr:
	content_letter_close(&lr);
	pr:
	content_proc_kill(&pr);
	return rv;
}

static int
command_thread(struct letter *letter, struct command_args *args)
{
	struct mailbox_thread thread;
	struct letter *let;

	mailbox_thread_init(args->mailbox, &thread, letter);

	while ((let = mailbox_thread_next(args->mailbox, &thread)) != NULL) {
		size_t idx;

		idx = let - args->mailbox->letters;
		if (letter_print(idx + 1, let) == -1)
			return -1;
	}

	return 0;
}

static int
command_unread(struct letter *letter, struct command_args *args)
{
	return command_flag(letter, args, 'S', 0);
}

static int
content_proc_ex_ignore(struct content_proc *pr,
		       const struct config_ignore *ignore)
{
	size_t i;

	for (i = 0; i < ignore->nheader; i++)
		if (content_proc_ignore(pr, ignore->headers[i],
					ignore->retain) == -1)
			return -1;
	return 0;
}

static int
letter_print(size_t nth, struct letter *letter)
{
	struct tm tm;
	char date[33];
	const char *subject;

	if (localtime_r(&letter->date, &tm) == NULL)
		return -1;
	if (strftime(date, sizeof(date), "%a %b %d %H:%M %Y", &tm) == 0)
		return -1;

	if ((subject = letter->subject) == NULL)
		subject = "No Subject";

	if (printf("%4zu %-24s %-32s %-30s\n", nth, date,
		   letter->from, subject) < 0)
		return -1;
	return 0;
}

static void
read_letters(const char *maildir, int ocur, int view_all,
	     struct mailbox *mailbox)
{
	DIR *cur;
	struct content_proc pr;
	int curfd;

	if ((curfd = dup(ocur)) == -1)
		err(1, "dup");
	if (fcntl(curfd, F_SETFD, FD_CLOEXEC) == -1)
		err(1, "fcntl");
	if ((cur = fdopendir(curfd)) == NULL)
		err(1, "fdopendir");

	if (content_proc_init(&pr, PATH_MAILZ_CONTENT) == -1)
		errx(1, "content_proc_init");
	mailbox_init(mailbox);

	for (;;) {
		struct content_summary sm;
		struct letter letter;
		struct dirent *de;
		int fd;

		errno = 0;
		if ((de = readdir(cur)) == NULL) {
			if (errno == 0)
				break;
			err(1, "readdir");
		}

		if (!strcmp(de->d_name, ".") || !strcmp(de->d_name, ".."))
			continue;

		if (!view_all && maildir_get_flag(de->d_name, 'S'))
			continue;

		if ((fd = openat(curfd, de->d_name, O_RDONLY | O_CLOEXEC)) == -1)
			err(1, "%s/cur/%s", maildir, de->d_name);
		if (content_proc_summary(&pr, &sm, fd) == -1)
			errx(1, "content_proc_summary: %s/cur/%s", maildir, de->d_name);

		letter.date = sm.date;
		letter.from = sm.from;
		letter.path = de->d_name;
		letter.subject = sm.have_subject ? sm.subject : NULL;

		if (mailbox_add_letter(mailbox, &letter) == -1)
			err(1, NULL); /* errno == ENOMEM */
	}

	mailbox_sort(mailbox);

	content_proc_kill(&pr);
	closedir(cur);
}

static void
setup_letters(const char *maildir, int root, int cur)
{
	DIR *new;
	int newfd;

	if ((newfd = openat(root, "new",
			    O_RDONLY | O_DIRECTORY | O_CLOEXEC)) == -1)
		err(1, "%s/new", maildir);
	if ((new = fdopendir(newfd)) == NULL)
		err(1, "fdopendir");

	for (;;) {
		char name[NAME_MAX + 1], *namep;
		struct dirent *de;

		errno = 0;

		if ((de = readdir(new)) == NULL) {
			if (errno == 0)
				break;
			err(1, "readdir");
		}

		if (!strcmp(de->d_name, ".") || !strcmp(de->d_name, ".."))
			continue;

		if (strchr(de->d_name, ':') == NULL) {
			int n;

			n = snprintf(name, sizeof(name), "%s:2,", de->d_name);
			if (n < 0 || (size_t)n >= sizeof(name))
				errc(1, ENAMETOOLONG, "rename %s/new/%s to %s/cur/%s:2,",
				     maildir, de->d_name, maildir, de->d_name);
			namep = name;
		}
		else
			namep = de->d_name;

		if (renameat(newfd, de->d_name, cur, namep) == -1)
			err(1, "rename %s/new/%s to %s/cur/%s",
			     maildir, de->d_name, maildir, namep);
	}

	closedir(new);
}

static void
usage(void)
{
	fprintf(stderr, "usage: mailz [-an] [-f file] mailbox\n");
	exit(2);
}

int
main(int argc, char *argv[])
{
	char *home, *slash, tmpdir[PATH_MAX];
	const char *address, *confpath, *maildir;
	struct config conf;
	struct config_mailbox *conf_mailbox;
	struct mailbox mailbox;
	int ch, cur, dryrun, n, root, view_all;

	confpath = NULL;
	dryrun = 0;
	view_all = 0;
	while ((ch = getopt(argc, argv, "af:n")) != -1) {
		switch (ch) {
		case 'a':
			view_all = 1;
			break;
		case 'f':
			confpath = optarg;
			break;
		case 'n':
			dryrun = 1;
			break;
		default:
			usage();
		}
	}

	argc -= optind;
	argv += optind;

	if (dryrun) {
		if (argc != 0)
			usage();
		parse_config(&conf, confpath);
		printf("configuration OK\n");
		return 0;
	}

	/*
	 * Delete trailing slash to make error messages nicer.
	 */
	if ((slash = strrchr(argv[0], '/')) != NULL && slash[1] == '\0')
		*slash = '\0';

	if (argc != 1)
		usage();

	if (setlocale(LC_CTYPE, "C.UTF-8") == NULL)
		errx(1, "setlocale");
	signal(SIGPIPE, SIG_IGN);

	parse_config(&conf, confpath);
	if ((conf_mailbox = config_mailbox(&conf, argv[0])) != NULL) {
		if (strlen(conf_mailbox->address) != 0)
			address = conf_mailbox->address;
		else
			address = conf.address;
		maildir = conf_mailbox->maildir;
	}
	else {
		address = conf.address;
		maildir = argv[0];
	}

	if ((root = open(maildir, O_RDONLY | O_DIRECTORY | O_CLOEXEC)) == -1)
		err(1, "%s", maildir);
	if ((cur = openat(root, "cur", O_RDONLY | O_DIRECTORY | O_CLOEXEC)) == -1)
		err(1, "%s/cur", maildir);

	if ((home = getenv("HOME")) == NULL)
		errx(1, "HOME not set");

	n = snprintf(tmpdir, sizeof(tmpdir), "%s/.mailz", home);
	if (n < 0 || (size_t)n >= sizeof(tmpdir))
		errx(1, "snprintf overflow due to large HOME");
	if (mkdir(tmpdir, 0700) == -1 && errno != EEXIST)
		err(1, "%s", tmpdir);

	if (unveil(tmpdir, "rwc") == -1)
		err(1, "%s", tmpdir);
	if (unveil(maildir, "rc") == -1)
		err(1, "%s", maildir);
	if (unveil(PATH_LESS, "x") == -1)
		err(1, "%s", PATH_LESS);
	if (unveil(PATH_MAILZ_CONTENT, "x") == -1)
		err(1, "%s", PATH_MAILZ_CONTENT);
	if (unveil(PATH_SENDMAIL, "x") == -1)
		err(1, "%s", PATH_SENDMAIL);
	if (pledge("stdio rpath wpath cpath sendfd proc exec", NULL) == -1)
		err(1, "pledge");

	setup_letters(maildir, root, cur);
	read_letters(maildir, cur, view_all, &mailbox);

	if (mailbox.nletter == 0)
		puts("No mail.");
	else {
		struct command_args args;
		size_t i;

		for (i = 0; i < mailbox.nletter; i++)
			letter_print(i + 1, &mailbox.letters[i]);

		args.addr = address;
		args.cur = cur;
		args.ignore = &conf.ignore;
		args.mailbox = &mailbox;
		args.maildir = maildir;
		args.tmpdir = tmpdir;

		commands_run(&args);
	}

	mailbox_free(&mailbox);
	rmdir(tmpdir);
	close(cur);
	close(root);
	config_free(&conf);
}
